### PR TITLE
Return value by parameter from scalars (v1)

### PR DESCRIPTION
**Pull request by losipiuk**
@arhimondr, @fiedukow 
Please give it first read.

### CODE REVIEW

**Comment by losipiuk** [comment]
ping
@kokosing, @pnowojski, anybody?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 1]
```diff
@@ -37,11 +41,48 @@ public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArgu
...
+        this.returnValueSliceLength = returnValueSliceLength;
+
+        int expectedParametersCount = methodHandle.type().parameterCount();
+        if (returnValueAsParameter) {
+            expectedParametersCount--;
+        }
+        if (instanceFactory.isPresent()) {
+            expectedParametersCount--;
+        }
+        if (methodHandle.type().parameterCount() >= 1 && methodHandle.type().parameterType(0) == ConnectorSession.class) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java]: As far as I understand we don't force any other "special" parameter to be on fixed position. Maybe we can break this requirement here as well (or force this requirement on other special arguments).

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 2]
```diff
@@ -37,11 +41,48 @@ public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArgu
...
+            expectedParametersCount--;
+        }
+        checkArgument(nullableArguments.size() == expectedParametersCount, "method handle parameters count mismatch");
+
+        if (returnValueAsParameter) {
+            checkArgument(!nullable, "nullable not supported for functions returning value as parameter");
+            checkArgument(getMethodHandle().type().returnType() == void.class, "methodHandle returns non-void for function returning value as parameter");
+            int returnParameterPos = instanceFactory.isPresent() ? 1 : 0;
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java:104]: I like the fact that we force position of special parameters but I'm not sure about forcing the order of them.
Even if we force the order - I think this is invalid when there is both instanceFactory and ConnectorSession passed? If that is correct please add test to cover that for future as well. Also if we want to force the order it would be nice to check it explicitly.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 3]
```diff
@@ -37,11 +41,48 @@ public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArgu
...
+        }
+        if (instanceFactory.isPresent()) {
+            expectedParametersCount--;
+        }
+        if (methodHandle.type().parameterCount() >= 1 && methodHandle.type().parameterType(0) == ConnectorSession.class) {
+            expectedParametersCount--;
+        }
+        checkArgument(nullableArguments.size() == expectedParametersCount, "method handle parameters count mismatch");
+
+        if (returnValueAsParameter) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java:101]: extract to `verifyReturnParameter` method?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 4]
```diff
@@ -286,12 +310,29 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
...
+    }
+
     private static final class ImplementationParser
     {
         private final String functionName;
-        private final boolean nullable;
+        private boolean nullable;
         private final List<Boolean> nullableArguments = new ArrayList<>();
-        private final String returnType;
+        private boolean returnAsParameter;
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:333]: final?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 5]
```diff
@@ -286,12 +310,29 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
...
+        private Optional<Integer> returnTypeSliceLength = Optional.empty();
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:334]: final?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 6]
```diff
@@ -286,12 +310,29 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
...
+        private String returnType;
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:335]: final?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 7]
```diff
@@ -286,12 +310,29 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
...
+
+    private static <T extends Annotation> Optional<T> getAnnotation(Method method, Class<T> annotationClass)
+    {
+        return getAnnotation(method.getDeclaredAnnotations(), annotationClass);
+    }
+
     private static final class ImplementationParser
     {
         private final String functionName;
-        private final boolean nullable;
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:292]: why?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 8]
```diff
@@ -353,13 +401,28 @@ private void parseArguments(Method method)
...
                     checkArgument(annotations.length == 1, "Meta parameters may only have a single annotation");
-                    checkArgument(argumentTypes.isEmpty(), "Meta parameter must come before parameters");
+                    checkArgument(argumentTypes.isEmpty() && !returnAsParameter, "Meta parameter must come before parameters");
                     Annotation annotation = annotations[0];
                     if (annotation instanceof TypeParameter) {
                         checkArgument(typeParameters.contains(annotation), "Injected type parameters must be declared with @TypeParameter annotation on the method");
                     }
                     dependencies.add(parseDependency(annotation));
                 }
+                else if (containsReturnParameter(annotations)) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:411]: What if it contains both return and meta parameter?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 9]
```diff
@@ -1089,6 +1091,11 @@ public static Object invoke(ConnectorSession session, ScalarFunctionImplementati
         if (handle.type().parameterCount() > 0 && handle.type().parameterType(0) == ConnectorSession.class) {
             handle = handle.bindTo(session);
         }
+        Optional<Slice> returnValue = Optional.empty();
+        if (function.isReturnValueAsParameter()) {
+            Preconditions.checkArgument(handle.type().parameterType(0) == Slice.class, "only Slice supported as return parameter");
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:1097]: ConnectorSession + ReturnValueAsParameter case?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 10]
```diff
@@ -399,6 +399,7 @@ public WindowFunctionSupplier load(SpecializedFunctionKey key)
                 .scalar(MapCardinalityFunction.class)
                 .scalar(MapConcatFunction.class)
                 .scalar(MapToMapCast.class)
+                .scalar(StringFunctions.FirstFiveCharsReversed.class)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:402]: Should we add test function to global FunctionRegistry?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 11]
```diff
@@ -406,9 +409,9 @@ public static Slice upper(@SqlType("varchar(x)") Slice slice)
     private static Slice pad(Slice text, long targetLength, Slice padString, int paddingOffset)
     {
         checkCondition(
-            0 <= targetLength && targetLength <= Integer.MAX_VALUE,
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java:409]: Unrelated changes.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 12]
```diff
@@ -124,4 +149,80 @@ public boolean isReturnValueAsParameter()
...
+        MethodType originalMethodType = originalMethodHandle.type();
+        CallSiteBinder callSiteBinder = new CallSiteBinder();
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName("ScalarImplementation"),
+                type(Object.class));
+
+        List<Parameter> newParameters = newArrayList();
+        for (int i = 0; i < originalMethodType.parameterCount(); ++i) {
+            if (i != (usesInstance ? 1 : 0)) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java:183]: What about session param?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 13]
```diff
@@ -124,4 +149,80 @@ public boolean isReturnValueAsParameter()
...
+
+        methodBody.append(constantInt(returnValueSliceLength))
+                .invokeStatic(Slices.class, "allocate", Slice.class, int.class)
+                .putVariable(returnVariable);
+
+        if (usesInstance) {
+            methodBody.append(newParameters.get(0));
+        }
+        methodBody.getVariable(returnVariable);
+        for (int i = usesInstance ? 1 : 0; i < newParameters.size(); ++i) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java:200]: ditto.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java [hunk 14]
```diff
@@ -124,4 +149,80 @@ public boolean isReturnValueAsParameter()
...
+            methodBody.append(newParameters.get(i));
+        }
+        methodBody.append(invoke(callSiteBinder.bind(originalMethodHandle), "originalMethod"));
+        methodBody.getVariable(returnVariable);
+        methodBody.ret(Slice.class);
+
+        Class<?> generatedClass = defineClass(classDefinition, Object.class, callSiteBinder.getBindings(), getClass().getClassLoader());
+        return lookupNewMethodHandle(generatedClass);
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java:208]: Isn't it possible that generated classes will have name conflicts with each other?

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java [hunk 15]
```diff
@@ -194,7 +195,7 @@ public void smokedTest()
         assertExecute("bound_pattern", VARCHAR, "%el%");
         assertExecute("bound_null_string", VARCHAR, null);
         assertExecute("bound_timestamp_with_timezone", TIMESTAMP_WITH_TIME_ZONE, new SqlTimestampWithTimeZone(new DateTime(1970, 1, 1, 0, 1, 0, 999, DateTimeZone.UTC).getMillis(), TimeZoneKey.getTimeZoneKey("Z")));
-        assertExecute("bound_binary_literal", VARBINARY, new SqlVarbinary(new byte[]{(byte) 0xab}));
+        assertExecute("bound_binary_literal", VARBINARY, new SqlVarbinary(new byte[] {(byte) 0xab}));
```
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:198]: unrelated change

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java [hunk 16]
```diff
@@ -239,7 +240,7 @@ public void testUnaryOperators()
         }
 
         for (Integer value : intLefts) {
-            Long longValue = value == null ? null : value *  10000000000L;
+            Long longValue = value == null ? null : value * 10000000000L;
```
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:243]: unrelated change

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java [hunk 17]
```diff
@@ -313,7 +314,7 @@ public void testBinaryOperatorsIntegralIntegral()
                 assertExecute(generateExpression("%s / %s", left, right), INTEGER, left == null || right == null ? null : left / right);
                 assertExecute(generateExpression("%s %% %s", left, right), INTEGER, left == null || right == null ? null : left % right);
 
-                Long longLeft = left == null ? null : left *  1000000000L;
+                Long longLeft = left == null ? null : left * 1000000000L;
```
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:317]: unrelated change

**Comment by fiedukow** [comment]
LGTM % comments
Also I think we should consider keeping return value in function using parameter as buffer to simplify ensureReturnAsReturn. We can take Slice as argument and return same slice instead of returning void and packing whole thing in wrapper that will return this Slice anyway.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java (hunk 4 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:333]: can't be as this is not set in constructor but in method called from constructor.
This is not nice but it is hard to restructure this code to have this fields final and have nice looking initialization.

Probably with major rewrite of this class it would be possible. I wan not focusing on that in this PR. I can clean this up but I will leave it until I fix issues in gist of this PR.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java (hunk 8 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:411]:
```
                    checkArgument(annotations.length == 1, "Meta parameters may only have a single annotation");
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 9 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:1097]: It is supported. Note that above we bind session as first parameter and overwrite `handle`.
And then we operate on new `handle` with session already bound.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 10 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:402]: We should not. But there is not other option for now AFAIK. Function list is hardcoded in FR constructor.
Later we can replace this `FirstFiveCharsReversed` with some real function (e.g. one of decimal operators)
